### DISCUSSION TITLE
How best to use God Ray effect off-axis?

### DISCUSSION THREAD

**Discussion by AidanNelson** [pmndrs/postprocessing#435] [Q&A] [answered]
Hello! 👋

I'm wondering how to achieve God Ray effects when the "sun" source is not directly within the camera's frame.  When I use this effect, my goal is to achieve a subtle overlay over the entire canvas, not only when the camera is directed towards the sun.  When the camera is pointed directly at the sun, the sun source washes out the view of the scene behind it.  My goal is to achieve the look of 'volumetric fog' rather than sun flare on a camera lens.

---

With [threejs' built-in post-processing God Rays effect](https://threejs.org/examples/?q=god#webgl_postprocessing_godrays) , I am able to see the effect even when the sun source is outside of the camera view:
![Screen Shot 2022-12-04 at 11 10 52 AM](https://user-images.githubusercontent.com/6486359/205502287-b343a0fb-ee40-4ab8-ac5c-537da9a64e2c.png)

With [postprocessings' God Ray effect](https://pmndrs.github.io/postprocessing/public/demo/#god-rays), the effect is only visible with the sun is within the frame of the camera.  
![Screen Shot 2022-12-04 at 11 13 45 AM](https://user-images.githubusercontent.com/6486359/205502230-3ad1576d-fdc9-4636-9ec2-42f697007259.gif)

Are there settings I could adjust within the postprocessing God Rays effect to achieve the look from the threejs God Rays?  Or best-practices for creating a scene where the sun is not as present, but the effect is?

**Comment by vanruesc** [comment, marked as answer]
The `GodRaysEffect` works by blurring the light source, so if the source isn't on screen, there's nothing to blur. The version in the three.js examples blurs the depth buffer and uses that as the light source. I'm considering moving to that approach with the next major release, but there are pros and cons for both approaches.

**Reply by AidanNelson** [reply]
Thanks for your answer!

If it doesn't complicate the API too much, I would argue for including both options.  The off-screen (threejs postprocessing) version provides a really nice subtle effect, which is especially useful in an environment which mixes 2D and 3D content (as opposed to a more fully immersive 3D environment).  With @react-three/fiber ecosystem growing, I think we'll see more of these use-cases.
